src/java.base/share/classes/java/math/BigDecimal.java

```diff
@@ -5219,55 +5215,118 @@ private static boolean needIncrement(MutableBigInteger mdivisor, int roundingMod
         return commonNeedIncrement(roundingMode, qsign, cmpFracHalf, mq.isOdd());
     }
 
+    /**
+     * {@code FIVE_TO_2_TO[n] == 5^(2^n)}
+     */
+    private static final BigInteger[] FIVE_TO_2_TO = new BigInteger[20];
+
+    static {
+        BigInteger pow = FIVE_TO_2_TO[0] = BigInteger.valueOf(5L);
+        for (int i = 1; i < FIVE_TO_2_TO.length; i++)
+            FIVE_TO_2_TO[i] = pow = pow.multiply(pow);
+    }
+
+    /**
+     * @param n a non-negative integer
+     * @return {@code 5^(2^n)}
+     */
+    private static BigInteger fiveToTwoToThe(int n) {
+        int i = Math.min(n, FIVE_TO_2_TO.length - 1);
+        BigInteger pow = FIVE_TO_2_TO[i];
+        for (; i < n; i++)
+            pow = pow.multiply(pow);
+
+        return pow;
+    }
+
+    private static final double LOG_5_OF_2 = Math.log(2.0) / Math.log(5.0);
+
     /**
      * Remove insignificant trailing zeros from this
      * {@code BigInteger} value until the preferred scale is reached or no
      * more zeros can be removed.  If the preferred scale is less than
      * Integer.MIN_VALUE, all the trailing zeros will be removed.
+     * Assumes {@code intVal != 0}.
      *
      * @return new {@code BigDecimal} with a scale possibly reduced
      * to be closed to the preferred scale.
      * @throws ArithmeticException if scale overflows.
      */
     private static BigDecimal createAndStripZerosToMatchScale(BigInteger intVal, int scale, long preferredScale) {
+        // avoid overflow of scale - preferredScale
+        preferredScale = Math.clamp(preferredScale, Integer.MIN_VALUE - 1L, Integer.MAX_VALUE);
+        // a multiple of 10^n must be a multiple of 2^n
+        int powsOf2 = intVal.getLowestSetBit();
+        long remainingZeros = Math.min(scale - preferredScale, powsOf2);
+        if (remainingZeros <= 0L)
+            return valueOf(intVal, scale, 0);
+
+        final int sign = intVal.signum;
+        if (sign < 0)
+            intVal = intVal.negate(); // speed up computation of shiftRight() and bitLength()
+
+        intVal = intVal.shiftRight(powsOf2); // remove powers of 2
+        // maxPowsOf5 == ceil(log5(intVal)) roughly
+        long maxPowsOf5 = (long) Math.ceil(intVal.bitLength() * LOG_5_OF_2);
```

Comment:
Please document the meaning of this variable in form of inequalities w.r.t. `intVal`.
Also, shouldn't it be named `maxExpOf5` or similar? It's an exponent, after all.

Reply:
The name `maxPowsOf5` has been chosen for consistence with `powsOf2`.

Reply:
Could make sense using Math.round() instead of Math.ceil() to get better upper bound?

Reply:
I'll have to check...
... tomorrow ;-)

Reply:
IIUC, the code assumes that the floating-point computation `Math.ceil(intVal.bitLength() * LOG_5_OF_2)` has the same value as the mathematical &lceil; `intVal.bitLength()` &times; log5(2) &rceil;.
I don't think this is safe, as it might happen that the computed and mathematical values are off by &plusmn;1.
To ensure 5^`maxPowsOf5` >= `intVal` (that is, maxPowsOf5 >= log5(intVal)) it would be more prudent to have
```
long maxPowsOf5 = (long) Math.ceil(intVal.bitLength() * LOG_5_OF_2) + 1;
```

But I think what you really want is maybe to meet 5^`maxPowsOf5` <= `intVal` < 5^(`maxPowsOf5` + 1) instead?

Reply:
Actually, if we reason in terms of "ulp vs precision", the computation should be safe: `Math.log()`'s results are within 1 ulp of the exact result, and the floating point operations are a multiplication and a division. The division to compute `LOG_5_OF_2` costs 1/2 ulp plus the errors of the operands, so 2.5 ulps. Same for multiplication, but `intVal.bitLength()` has an exact `double` value, so the total roundoff error of `intVal.bitLength() * LOG_5_OF_2` is 3 ulps. Since `intVal.bitLength() * LOG_5_OF_2` representable with 31 bits, and double has 53 bits for the significand, we can reasonably say that `Math.ceil()` can always guarantee `maxPowsOf5 >= log5(intVal)`.

Reply:
If the mathematical value v of the product and its floating-point value fp are separated by an integer i in such a way that fp < i < v, we are in trouble: the ceilings will be different, even if the values are very close to each other.

Reply:
One might prove that this cannot happen for a specific approximation of log5(2), like `LOG_5_OF_2` and for all bit length, but I don't think it is worthwhile to put too much effort on this, given the performance figures.

Reply:
> If the mathematical value v of the product and its floating-point value fp are separated by an integer i in such a way that fp < i < v, we are in trouble: the ceilings will be different, even if the values are very close to each other.

In that unlikely case, we would have `maxPowsOf5 >= floor(log5(intVal))`, which is still acceptable as an upper bound.

Reply:
And considering that `Math.round()` rounds to the closest integer, it also should assure `maxPowsOf5 >= floor(log5(intVal))`, giving a better upper bound.

Reply:
As I asked above, what's the property that should hold?
5^`maxPowsOf5` <= `intVal` < 5^(`maxPowsOf5` + 1)?

Reply:
Sorry, this is not a good question, disregard.

Reply:
> As I asked above, what's the property that should hold? 5^`maxPowsOf5` <= `intVal` < 5^(`maxPowsOf5` + 1)?

Yes, as `floor(log5(intVal)) = max{integer n : 5^n <= intVal}`.

Reply:
This doesn't say much about `maxPowsOf5`, though. You are not using `intVal` but 2^`bitLength` > `intVal` in the computation of `maxPowsOf5`.
So maybe the property you are looking for is
`maxPowsOf5` - 1 <= log5(intVal) < `maxPowsOf5` || `maxPowsOf5` <= log5(intVal) < `maxPowsOf5` + 1

Reply:
> This doesn't say much about `maxPowsOf5`, though. You are not using `intVal` but 2^`bitLength` > `intVal` in the computation of `maxPowsOf5`. So maybe the property you are looking for is `maxPowsOf5` - 1 <= log5(intVal) < `maxPowsOf5` || `maxPowsOf5` <= log5(intVal) < `maxPowsOf5` + 1

Actually, it is sufficient that floor(log5(intVal)) >= max{integer n : 5^n <= intVal} to get a correct upper bound, not that are equal. But the closer they are, the better.

Reply:
> This doesn't say much about `maxPowsOf5`, though. You are not using `intVal` but 2^`bitLength` > `intVal` in the computation of `maxPowsOf5`. So maybe the property you are looking for is `maxPowsOf5` - 1 <= log5(intVal) < `maxPowsOf5` || `maxPowsOf5` <= log5(intVal) < `maxPowsOf5` + 1

To clarify as much as possible, the weakest property that `maxPowsOf5` must satisfy is:
`maxPowsOf5 >= max{integer n : intVal % 5^n == 0}`.
Clearly, it is true that `max{integer n : 5^n <= intVal} >= max{n : intVal % 5^n == 0}`;
and since `floor(log5(intVal)) == max{integer n : 5^n <= intVal}`,
it follows that `maxPowsOf5 >= floor(log5(intVal)) ==> maxPowsOf5 >= max{integer n : intVal % 5^n == 0}.`